airflow/providers/amazon/aws/hooks/elasticache.py

```diff
@@ -0,0 +1,238 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import re
+from time import sleep
+
+from airflow.exceptions import AirflowException
+from airflow.providers.amazon.aws.hooks.base_aws import AwsBaseHook
+
+
+class ElastiCacheDescribeResponseKeys:
+    """
+    Constants for ElastiCache describe API response keys
+    """
+    REPLICATION_GROUPS = 'ReplicationGroups'
+    REPLICATION_GROUP = 'ReplicationGroup'
+    STATUS = 'Status'
+
+
+class ElastiCacheDeleteResponses:
+    """
+    Constants for ElastiCache delete API response messages
+    """
+    RESPONSE_NOT_VALID_STATE = 'not in a valid state to be deleted'
+    RESPONSE_DELETING_STATE = 'has status deleting which is not valid for deletion'
+    RESPONSE_MODIFYING_STATE = 'has status modifying which is not valid for deletion'
+
+
+class ElastiCacheStatus:
+    """
+    Constants for ElastiCache replication group status
+    """
+    STATUS_CREATING = 'creating'
+    STATUS_AVAILABLE = 'available'
+    STATUS_DELETING = 'deleting'
+    STATUS_CREATE_FAILED = 'create-failed'
+
+
+class ElastiCacheHook(AwsBaseHook):
+    """
+    Interact with AWS ElastiCache
+    """
+
+    def __init__(
+        self, max_retries=10, exponential_back_off_factor=1, initial_poke_interval=60, *args, **kwargs
+    ):
+        """
+        :param max_retries: Max tries for checking availability of and deleting replication group
+        :param exponential_back_off_factor: Factor for deciding next sleep time
+        :param initial_poke_interval: Initial sleep time
+        """
+        self.max_retries = max_retries
+        self.exponential_back_off_factor = exponential_back_off_factor
+        self.initial_poke_interval = initial_poke_interval
+
+        super().__init__(client_type='elasticache', *args, **kwargs)
+
+    def __create_replication_group(self, config):
```

Comment:
why are these private?

Reply:
different question: why are they even separate methods?

Reply:
My thought behind this is: this method (private) just does 1 thing which is makes call to the 3rd party service and returns the response as is, no other decoration of any sort.

Then we have a wrapper on top of this which can do a lot of things on top of the response. In this case it is just checking on `status` of the replication group. 

But if we look at `delete_replication_group` API, it just a wrapper around `_delete_replication_group` which does all the work, retries and all, but in turn calls `__delete_replication_group` to make the 3rd party API call.

So keeping separate method I think helps to write more manageable and simple code, following Single Responsibility Principle.

But I do agree with your point that why are these private. My initial thought was, clients just use the public APIs and those handle everything, but maybe that thinking is not correct. Please provide your thoughts on this.